pennylane/tape/__init__.py

```diff
@@ -40,6 +41,128 @@
+    applied without requiring a QNode.
+
+    **Example**:
+
+    The OperationRecorder is a context manager. Executing templates
+    or quantum functions stores resulting applied operators in the
+    recorder, which can then be printed.
+
+    >>> weights = qml.init.strong_ent_layers_normal(n_layers=1, n_wires=2)
+    >>>
+    >>> with OperationRecorder() as rec:
+    >>>    qml.templates.layers.StronglyEntanglingLayers(*weights, wires=[0, 1])
+    >>>
+    >>> print(rec)
+    Operations
+    ==========
+    Rot(-0.10832656163640327, 0.14429091013664083, -0.010835826725765343, wires=[0])
+    Rot(-0.11254523669444501, 0.0947222564914006, -0.09139600968423377, wires=[1])
+    CNOT(wires=[0, 1])
+    CNOT(wires=[1, 0])
+
+    Alternatively, the :attr:`~.OperationRecorder.queue` attribute can be used
+    to directly access the applied :class:`~.Operation` and :class:`~.Observable`
+    objects.
+
+    Attributes:
+        queue (List[Operator]): list of operators applied within
+            the OperatorRecorder context, includes operations and observables
+        operations (List[Operation]): list of operations applied within
+            the OperatorRecorder context
+        observables (List[Observable]): list of observables applied within
+            the OperatorRecorder context
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.ops = None
+        self.obs = None
+
+    def _process_queue(self):
+        super()._process_queue()
+
+        for obj, info in self._queue.items():
+            QueuingContext.append(obj, **info)
+
+        # remove the operation recorder from the queuing
+        # context
+        QueuingContext.remove(self)
+
+        new_tape = self.expand(depth=5, stop_at=lambda obj: not isinstance(obj, QuantumTape))
+        self.ops = new_tape.operations
+        self.obs = new_tape.observables
+
+    def __str__(self):
+        output = ""
+        output += "Operations\n"
+        output += "==========\n"
+        for op in self.ops:
+            output += repr(op) + "\n"
+
+        output += "\n"
+        output += "Observables\n"
+        output += "==========\n"
+        for op in self.obs:
+            output += repr(op) + "\n"
+
+        return output
+
+    @property
+    def queue(self):
+        return self.ops + self.obs
+
+
+def TapeTemplateDecorator(func):
+    """Register a quantum template with PennyLane.
+
+    This decorator wraps the given function and makes it return a list of all queued Operations.
+
+    **Example:**
+
+    When defining a :doc:`template </introduction/templates>`, simply decorate
+    the template function with this decorator.
+
+    .. code-block:: python3
+
+        @qml.template
+        def bell_state_preparation(wires):
+            qml.Hadamard(wires=wires[0])
+            qml.CNOT(wires=wires)
+
+    This registers the template with PennyLane, making it compatible with
+    functions that act on templates, such as :func:`pennylane.inv`:

```

Comment:
So with this PR the `inv` function _does_ work?

Reply:
Yep!